pkg/pillar/cmd/zedrouter/pubsubhandlers.go

```diff
@@ -629,3 +629,75 @@ func (z *zedrouter) handleAppInstDelete(ctxArg interface{}, key string,
 	z.unpublishAppInstMetadata(appInstMetadata)
 	z.log.Functionf("handleAppInstDelete(%s) done", key)
 }
+
+func (z *zedrouter) handlePatchEnvelopeCreate(ctxArg interface{}, key string,
+	configArg interface{}) {
+	peInfo := configArg.(types.PatchEnvelopeInfoList)
+	z.log.Functionf("handlePatchEnvelopeCreate: (UUID: %s) %v", key, peInfo.Envelopes)
+
+	z.patchEnvelopes.PatchEnvelopeInfoCh <- peInfo.Envelopes
```

Comment:
Can this block reading from the channel? If so, for how long?

Reply:
this is O(N) where N -- number of all volume references we have for all patch envelopes. This happens because we might have already information about volume statuses for volume references that we just got. Processing PatchEnvelopeInfoCh happens only only when zedagent's parseConfig is triggered, which can be qualified as not so frequent.
But we are processing both channels in same function, so during processing PatchEnvelopeInfoCh we will block writing to VolumeStatusCh. It might give us problems if number of patch envelopes will be big enough. We can make processing PatchEnvelopeInfoCh O(log N) if we will use RBT instead of array to link existing volume references with new []PatchEnvelopeInfo we got. Normally I'd go for hash table, but I'm not sure how performant it'll be in golang and RBT gives you predictable time of fetching data.
In addition we can always, create buffer for VolumeStatuses we get from VolumeStatusCh, because receiving patchEnvelopeInfo is rather rare.

Reply:
The way we avoid these issues about size of channels in general is to save the intended state (which is always bound by the number of objects) and then get notifications when there is a state change (and such notications can use a channel with a conditional send).

In some places the state is saved using pubsub and in other cases some datastructure in memory (such as a locked map).
That avoids issues with deadlocks or unbounded processing time in handlers due to using channels to send state changes.

Do you think such an approach makes sense here?

Reply:
I put a comment explaining how channels should be used and buffered them, I think this should be enough to avoid issues with deadlocks or unbounded processing time in handlers